### PR TITLE
Documentation for the Trainer API

### PR DESCRIPTION
**Pull request by sgugger**
This PR introduces documentation for the Trainer/TFTrainer classes and the argument dataclasses.

While documenting, I noticed that the method `Trainer.evaluate` and `TFTrainer.evaluate` had an argument `pred_loss_only` which was not used at all (my guess is that it's the same argument passed at init that is used). I removed that argument. I'm aware this is a breaking change but, as a user, I'd expect a hard error when passing something that is not used at all. Let me know if there was a good reason for it.

### CODE REVIEW

File: src/transformers/trainer_tf.py [hunk 1]
```diff
@@ -78,6 +104,9 @@ def __init__(
         set_seed(self.args.seed)
 
     def get_train_tfdataset(self) -> tf.data.Dataset:
+        """
+        Returns the training :class:`~tf.data.Datasetr`.
```
Comment (jplu) [inline src/transformers/trainer_tf.py:108]: Small typo `tf.data.Datasetr` -> `tf.data.Dataset`

File: src/transformers/trainer_tf.py [hunk 2]
```diff
@@ -29,6 +29,32 @@ def set_seed(seed: int):
...
+            The function that will be used to compute metrics at evaluation. Must take a
+            :class:`~transformers.EvalPrediction` and return a dictionary string to metric values.
+        prediction_loss_only (:obj:`bool`, `optional`, defaults to `False`):
+            When performing evaluation and predictions, only returns the loss.
+        tb_writer (:obj:`SummaryWriter`, `optional`):
+            Object to write to TensorBoard.
+        optimizers (:obj:`Tuple[tf.keras.optimizers.Optimizer, tf.keras.optimizers.schedules.LearningRateSchedule]`, `optional`):
+            A tuple containing the optimizer and the scheduler to use. Will default to an instance of
+            :class:`~transformers.AdamW` on your model and a scheduler given by
+            :func:`~transformers.get_linear_schedule_with_warmup` controlled by :obj:`args`.
```
Comment (jplu) [inline src/transformers/trainer_tf.py:55]: These two have to be updated:
:class: `tf.keras.optimizers.Adam` if the `weight_decay_rate` argument equals to 0 else `transformers.AdamWeightDecay`
:func: `tf.keras.optimizers.schedules.PolynomialDecay` if the `num_warmup_steps` argument equals to 0 else `transformers.WarmUp`

File: src/transformers/trainer_tf.py [hunk 3]
```diff
@@ -29,6 +29,32 @@ def set_seed(seed: int):
...
+        train_dataset (:class:`~tf.data.Dataset`, `optional`):
+            The dataset to use for training.
+        eval_dataset (:class:`~tf.data.Dataset`, `optional`):
+            The dataset to use for evaluation.
+        compute_metrics (:obj:`Callable[[EvalPrediction], Dict]`, `optional`):
+            The function that will be used to compute metrics at evaluation. Must take a
+            :class:`~transformers.EvalPrediction` and return a dictionary string to metric values.
+        prediction_loss_only (:obj:`bool`, `optional`, defaults to `False`):
+            When performing evaluation and predictions, only returns the loss.
+        tb_writer (:obj:`SummaryWriter`, `optional`):
```
Comment (jplu) [inline src/transformers/trainer_tf.py:50]: `SummaryWriter` -> `tf.summary.SummaryWriter`

**Comment by julien-c** [comment]
Great! Much needed! Also adding @thomwolf who might want to chime in

File: src/transformers/training_args_tf.py [hunk 4]
```diff
@@ -59,9 +138,15 @@ def _setup_strategy(self) -> Tuple["tf.distribute.Strategy", int]:
...
+        """
+        The strategy used for distributed training.
+        """
         return self._setup_strategy
 
     @property
     @tf_required
     def n_gpu(self) -> int:
+        """
+        The number of GPUs used in this training.
```
Comment (jplu) [inline src/transformers/training_args_tf.py:150]: The number of replicas, not GPUs (means CPU/TPU cores as well). I think we should think to rename that function to `n_replica` or `n_device`. @julien-c WDYT?

**Comment by jplu** [comment]
The TF documentation looks much better!! :smile: Thanks @sgugger

**Comment by sgugger** [comment]
Thanks for catching all my bad copy-pastes @jplu 
Concerning `n_gpu`, I think the change of name to `n_replica`/`n_device` might be welcome, especially since it does not represent the same thing as `n_gpu` in the PyTorch `Trainer` (which is only more than 1 when using dp instead of ddp). But that probably would go better in another PR!

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -141,6 +205,9 @@ class TrainingArguments:
 
     @property
     def train_batch_size(self) -> int:
+        """
+        The actual batch size for training (may differ for :obj:`per_gpu_train_batch_size` in distributed training).
```
Comment (LysandreJik) [inline src/transformers/training_args.py:209]: may differ for or from?

File: src/transformers/training_args.py [hunk 6]
```diff
@@ -151,6 +218,9 @@ def train_batch_size(self) -> int:
 
     @property
     def eval_batch_size(self) -> int:
+        """
+        The actual batch size for evaluation (may differ for :obj:`per_gpu_eval_batch_size` in distributed training).
```
Comment (LysandreJik) [inline src/transformers/training_args.py:222]: may differ for or from?

File: src/transformers/training_args_tf.py [hunk 7]
```diff
@@ -59,9 +138,15 @@ def _setup_strategy(self) -> Tuple["tf.distribute.Strategy", int]:
...
+        """
+        The strategy used for distributed training.
+        """
         return self._setup_strategy
 
     @property
     @tf_required
     def n_gpu(self) -> int:
+        """
+        The number of replicas (GPUs or TPUs) used in this training.
```
Comment (LysandreJik) [inline src/transformers/training_args_tf.py:150]: I'd say TPU core here instead of TPUs

**Review by LysandreJik — APPROVED** [review]
Very cool!
